src/psbt/mod.rs

```diff
@@ -871,54 +964,64 @@ fn script_code_wpkh(script: &Script) -> Script {
     script_code.push(0xac);
     Script::from(script_code)
 }
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
+/// Return error type of [`PsbtInputExt::update_with_descriptor`]
+pub enum UpdateInputError {
+    /// Derivation error
+    DerivationError(descriptor::ConversionError),
+    /// `witness_utxo` check failed
+    WitnessUtxoCheck,
+    /// `non_witness_utxo` check failed
+    NonWitnessUtxoCheck,
```

Comment:
One of the `MissingUtxoError` is because it is not clear which field should be used for segwit transactions and how we should error it. 

The current implementation is finalizer is as follows:
```rust
// Get the spending utxo for this psbt input
pub(super) fn get_utxo(psbt: &Psbt, index: usize) -> Result<&bitcoin::TxOut, InputError> {
    let inp = &psbt.inputs[index];
    let utxo = if let Some(ref witness_utxo) = inp.witness_utxo {
        &witness_utxo
    } else if let Some(ref non_witness_utxo) = inp.non_witness_utxo {
        let vout = psbt.unsigned_tx.input[index].previous_output.vout;
        &non_witness_utxo.output[vout as usize]
    } else {
        return Err(InputError::MissingUtxo);
    };
    Ok(utxo)
}
```

Reply:
The lack of clarity about "which field should be used" is why I made a conscious design decision to avoid using the above logic and I move to a API where the caller is saying which fields should be checked against the descriptor within the context of the application. So the caller asserts that there should be `witness_utxo` and/or `non_witness_utxo` and the function checks that it is present and correct (at least when it is plausibly appropriate).

For example, using `get_utxo` the function would not check that `non_witness_utxo` is present even if the application requires it and even if it was present would not check that it has the correct `scriptPubkey`. Coldcard, for example, requires that `non_witness_utxo` is present correct even if `witness_utxo` is provided.

You could argue that it is not the responsibility of a function that updates the `psbt::Input` with stuff from a descriptor to enforce these policies but I think it's just way to convenient to do it here and creates a single PSBT input state transition where you can guarantee your input is correct after you call this function on it.